src/lib/backend/equinoxe_cohttp.ml

```diff
@@ -0,0 +1,148 @@
+
+open Lwt.Syntax
+open Lwt.Infix
+module Client = Cohttp_lwt_unix.Client
+module Json = Equinoxe.Json
+module Utils = Equinoxe.Private.Utils
+
+module Backend = struct
+  (**** Type definitions ****)
+
+  type t = { address : string; token : string }
+
+  let token t = t.token
+  let address t = t.address
+
+  (**** Default values ****)
+
+  let build_header token =
+    let token = if token = "" then [] else [ ("X-Auth-Token", token) ] in
+    token @ [ ("Content-Type", "application/json") ] |> Cohttp.Header.of_list
+
+  let equinoxe_default_path = Utils.Sys.path_from_home_dir ".config/equinoxe/"
+
+  (***** Helpers *****)
+
+  let compute ~time ~f =
+    let* status =
+      Lwt.pick
+        [
+          (f () >|= fun v -> `Done v);
+          (Lwt_unix.sleep time >|= fun () -> `Timeout);
+        ]
+    in
+    match status with
+    | `Done v -> Lwt.return v
+    | `Timeout -> Lwt.fail_with "Http request timeout"
+
+  let convert_to_json (resp, body) =
+    let code = Cohttp.(Response.status resp |> Code.code_of_status) in
+    if code >= 200 && code < 300 then
+      Lwt.catch
+        (fun () ->
+          let+ body = Cohttp_lwt.Body.to_string body in
+          match body with "" -> Json.of_string "{ }" | s -> Json.of_string s)
+        (fun e -> Lwt.return @@ Json.error (Printexc.to_string e))
+    else
+      let msg = Format.sprintf "Cohttp exits with HTTP code %d" code in
+      Lwt.return @@ Json.error msg
+
+  let token_from_path token_path =
+    match Utils.Reader.read_token_opt token_path with
+    | None -> raise Not_found
+    | Some token -> token
+
+  let default_token () =
+    let token_path = Filename.concat equinoxe_default_path "token" in
+    token_from_path token_path
+
+  let get_token = function
+    | `Default -> default_token ()
+    | `Path token_path -> token_from_path token_path
+    | `Str token -> token
+
+  (**** Http methode ****)
+
+  let compute = compute ~time:10.0
+
+  let get_from t path =
+    let headers = build_header t.token in
+    let url = Filename.concat t.address path |> Uri.of_string in
+    let f () = Client.get ~headers url in
+    compute ~f
+
+  let post_from t ~path body =
+    let headers = build_header t.token in
+    let url = Filename.concat t.address path |> Uri.of_string in
+    let body = Cohttp_lwt.Body.of_string body in
+    let f () = Client.post ~headers ~body url in
+    compute ~f
+
+  let put_from t ~path body =
+    let headers = build_header t.token in
+    let url = Filename.concat t.address path |> Uri.of_string in
+    let body = Cohttp_lwt.Body.of_string body in
+    let f () = Client.put ~headers ~body url in
+    compute ~f
+
+  let delete_from t path =
+    let headers = build_header t.token in
+    let url = Filename.concat t.address path |> Uri.of_string in
+    let f () = Client.delete ~headers url in
+    compute ~f
+
+  (**** API ****)
+
+  let create ~address ?(token = `Default) () =
+    let token = get_token token in
```

Comment:
The `~token` could be passed as a `string`, it would be easier to handle the three choices `Default|Path|Str` from `Equinoxe.create` than in each backend.

Also I'm not a fan of the `Default` that reads an undocumented file at `~/.config/equinoxe` !.. Perhaps an alternative would be to drop the variants in favor of three functions `token_of_string`, `token_from_file`, `token_from_my_home_config_directory`? (The last two are easily defined from the string one so I don't think people will miss them)